openhtf/io/output/__init__.py

```diff
@@ -1,26 +1,41 @@
-# Copyright 2014 Google Inc. All Rights Reserved.
+"""Output module."""
 
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-# You may obtain a copy of the License at
+import base64
+import os
+import tempfile
+from openhtf.util import data
 
-#     http://www.apache.org/licenses/LICENSE-2.0
 
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
+class OutputToFile(object):
+  """Class for formatting filename and writing to file."""
 
-"""Openhtf output modules."""
+  def __init__(self, filename_pattern=None):
+    self.filename_pattern = filename_pattern
 
-import openhtf.io.output.json_factory
-import openhtf.io.output.mfg_inspector
+  def __call__(self, test_record):
+    test_record_dict = self.ConvertToDict(test_record)
+    write_data = self.GetWriteData(test_record)
 
-# pylint: disable=invalid-name
+    if isinstance(self.filename_pattern, basestring):
+      filename = self.Format(test_record_dict)
+      self.Save(filename, write_data)
+      return filename
+    self.filename_pattern.write(write_data)
 
-# Classes used for output
-OutputToJSON = json_factory.OutputToJSON
-UploadToMfgInspector = mfg_inspector.UploadToMfgInspector
-OutputToTestRunProto = mfg_inspector.OutputToTestRunProto
-UploadOrOutput = mfg_inspector.UploadOrOutput
+  def ConvertToDict(self, test_record):
+    return data.ConvertToBaseTypes(test_record)
+
+  def GetWriteData(self, test_record):
+    return str(self.ConvertToDict(test_record))
+
+  def Format(self, test_record_dict):
```

Comment:
rather than handling just the `.format()` here, I'd also handle the possibility of a callable, that way all the 'handle possible inputs of filename_pattern' logic is in the same place.  something like:
```Python
  def open_output_file(self, test_record):
    record_dict = data.ConvertToBaseTypes(test_record)
    if isinstance(self.filename_pattern, basestring):
      if '{' in self.filename_pattern:
        # Replace open() with atomic file open.
        return open(self.filename_pattern.format(**record_dict), 'wb')
      elif '%' in self.filename_pattern:
        raise ValueError('%-style filename patterns deprecated, use .format() syntax')
      return open(self.filename_pattern, 'wb')
    if callable(self.filename_pattern):
      return open(self.filename_pattern(record_dict), 'wb')
    if hasattr(self.filename_pattern, 'write'):
      return self.filename_pattern
    raise ValueError('filename_pattern must be string, callable, or File-like object')

  def __call__(self, test_record):
    with self.open_output_file(test_record) as outfile:
      outfile.write(self.serialize_test_record(test_record))
```